src/api/spotify_client.py

```diff
@@ -1,54 +1,96 @@
+
+    def get_access_token(self) -> str | None:
+        """
+        Fetches an OAuth access token from Spotify's API using Client Credentials Flow.
+        Returns the access token if successful, otherwise returns None.
+        """
+
+        # Client ID and Client Secret are taken from class attributes
+        client_id: str = self.client_id
+        client_secret: str = self.client_secret
 
         # Spotify's Accounts service token URL
-        token_url = "https://accounts.spotify.com/api/token"
+        token_url: str = "https://accounts.spotify.com/api/token"
 
         # Encode the Client ID and Client Secret in base64 (as required by Spotify)
-        client_creds = f"{client_id}:{client_secret}"
-        encoded_creds = b64encode(client_creds.encode()).decode()
+        client_creds: str = f"{client_id}:{client_secret}"
+        encoded_creds: str = b64encode(client_creds.encode()).decode()
 
-        # Prepare the headers for the request
-        headers = {
+        # Prepare the headers for the request, including the encoded credentials
+        headers: dict[str, str] = {
             "Authorization": f"Basic {encoded_creds}",
-            "Content-Type": "application/x-www-form-urlencoded"
+            "Content-Type": "application/x-www-form-urlencoded",
         }
 
-        # Prepare the data to send in the POST request
-        data = {
-            "grant_type": "client_credentials"
-        }
+        # Prepare the data to send in the POST request, specifically the grant type
+        data: dict[str, str] = {"grant_type": "client_credentials"}
 
         # Make the POST request to get the access token
-        response = requests.post(token_url, headers=headers, data=data)
+        response: requests.Response = requests.post(token_url, headers=headers, data=data)
 
-        # Parse the response as JSON
+        # Parse the response and check if the request was successful
         if response.status_code == 200:
-            token_info = response.json()
-            access_token = token_info['access_token']
+            token_info: dict[str, Any] = response.json()
+            access_token: str = token_info.get("access_token", "")
             print(f"Access Token: {access_token}")
         else:
+            # Log error in case the request fails
             print(f"Failed to get access token: {response.status_code}")
             print(response.json())
             access_token = None
 
+        # Store access token in the class and return it
         self.access_token = access_token
         return access_token
+
+    def get_song_features(self, track_id: str) -> dict[str, Any]:
+        """
+        Fetches the audio features of a given Spotify track using its track ID.
+        Returns a dictionary containing the track's audio features.
+        """
+        
+        # Fetch the access token (if it's not available, make the request to get it)
+        access_token: str | None = self.get_access_token()
```

Comment:
Check is self.get_access_token already exists. If not exists then only create new one. Or if access token has expired. Will prevent many api calls.

Reply:
Also assign to self.access_token for use by other functions